File: omini_connect/portal/sdk/flutter/lib/ominiconnect.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// OminiConnect Flutter SDK.
/// Connect AI agents to external platforms via OAuth.
library ominiconnect;

/// Thrown for all OminiConnect errors.
class OminiConnectException implements Exception {
  final String message;
  final int? statusCode;
  final String? body;

  OminiConnectException(this.message, {this.statusCode, this.body});

  @override
  String toString() => 'OminiConnectException: $message';

  bool get isAuth => statusCode == 401;
  bool get isRateLimited => statusCode == 429;
  bool get isNotFound => statusCode == 404;
}

/// Raised when API key is invalid or missing.
class AuthException extends OminiConnectException {
  AuthException() : super('Invalid or missing API key', statusCode: 401);
}

/// Raised when a platform is not connected.
class ConnectorNotFoundException extends OminiConnectException {
  ConnectorNotFoundException(String platform)
      : super('Platform not connected: $platform', statusCode: 404);
}

/// Raised when a tool is not found.
class ToolNotFoundException extends OminiConnectException {
  ToolNotFoundException(String slug)
      : super('Tool not found: $slug', statusCode: 404);
}

/// Raised when scopes are insufficient.
class ScopeInsufficientException extends OminiConnectException {
  ScopeInsufficientException()
      : super('Insufficient scopes for this operation', statusCode: 403);
}

// ─── Types ──────────────────────────────────────────────────────────────────

/// A connected platform.
class Connector {
  final String platform;
  final bool enabled;
  final List<String> scopes;
  final DateTime createdAt;

  Connector({
    required this.platform,
    required this.enabled,
    required this.scopes,
    required this.createdAt,
  });

  factory Connector.fromJson(Map<String, dynamic> json) {
    return Connector(
      platform: json['platform'] as String,
      enabled: json['enabled'] as bool,
      scopes: List<String>.from(json['scopes'] ?? []),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// API key returned on creation — store securely, shown only once.
class ApiKeyCreated {
  final String key;
  final String label;
  final DateTime createdAt;

  ApiKeyCreated({
    required this.key,
    required this.label,
    required this.createdAt,
  });

  factory ApiKeyCreated.fromJson(Map<String, dynamic> json) {
    return ApiKeyCreated(
      key: json['key'] as String,
      label: json['label'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// API key summary — raw key is never returned.
class ApiKeySummary {
  final String keyHash;
  final String label;
  final DateTime createdAt;

  ApiKeySummary({
    required this.keyHash,
    required this.label,
    required this.createdAt,
  });

  factory ApiKeySummary.fromJson(Map<String, dynamic> json) {
    return ApiKeySummary(
      keyHash: json['key_hash'] as String,
      label: json['label'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// A tool from the registry.
class ToolSummary {
  final String slug;
  final String name;
  final String description;
  final String method;
  final String endpoint;
  final List<String> scopes;
  final String scopeSatisfied;
  final List<String> tags;

  ToolSummary({
    required this.slug,
    required this.name,
    required this.description,
    required this.method,
    required this.endpoint,
    required this.scopes,
    required this.scopeSatisfied,
    required this.tags,
  });

  factory ToolSummary.fromJson(Map<String, dynamic> json) {
    return ToolSummary(
      slug: json['slug'] as String,
      name: json['name'] as String,
      description: json['description'] as String,
      method: json['method'] as String,
      endpoint: json['endpoint'] as String,
      scopes: List<String>.from(json['scopes'] ?? []),
      scopeSatisfied: json['scope_satisfied'] as String,
      tags: List<String>.from(json['tags'] ?? []),
    );
  }
}

/// A toolkit grouping related tools.
class Toolkit {
  final String slug;
  final String name;
  final String? logo;
  final String provider;
  final List<ToolSummary> tools;

  Toolkit({
    required this.slug,
    required this.name,
    this.logo,
    required this.provider,
    required this.tools,
  });

  factory Toolkit.fromJson(Map<String, dynamic> json) {
    return Toolkit(
      slug: json['slug'] as String,
      name: json['name'] as String,
      logo: json['logo'] as String?,
      provider: json['provider'] as String,
      tools: (json['tools'] as List<dynamic>?)
              ?.map((t) => ToolSummary.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

/// Response from tools list.
class ToolkitsResponse {
  final List<Toolkit> toolkits;

  ToolkitsResponse({required this.toolkits});

  factory ToolkitsResponse.fromJson(Map<String, dynamic> json) {
    return ToolkitsResponse(
      toolkits: (json['toolkits'] as List<dynamic>?)
              ?.map((t) => Toolkit.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

/// Response from tools search.
class ToolsSearchResponse {
  final List<ToolSummary> tools;
  final String query;

  ToolsSearchResponse({required this.tools, required this.query});

  factory ToolsSearchResponse.fromJson(Map<String, dynamic> json) {
    return ToolsSearchResponse(
      tools: (json['tools'] as List<dynamic>?)
              ?.map((t) => ToolSummary.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
      query: json['query'] as String,
    );
  }
}

/// Result from tool execution.
class ToolExecuteResult {
  final bool ok;
  final dynamic body;
  final String? error;
  final String? callId;
  final String? status;
  final int? durationMs;

  ToolExecuteResult({
    required this.ok,
    this.body,
    this.error,
    this.callId,
    this.status,
    this.durationMs,
  });

  factory ToolExecuteResult.fromJson(Map<String, dynamic> json) {
    return ToolExecuteResult(
      ok: json['ok'] as bool,
      body: json['body'],
      error: json['error'] as String?,
      callId: json['call_id'] as String?,
      status: json['status'] as String?,
      durationMs: json['duration_ms'] as int?,
    );
  }
}

// ─── Client ─────────────────────────────────────────────────────────────────

class OminiConnect {
  final String apiKey;
  final String baseUrl;
  final http.Client _http;

  late final ConnectorsManager connectors;
  late final ToolsManager tools;
  late final ApiKeysManager apiKeys;

  OminiConnect({
    required this.apiKey,
    this.baseUrl = 'http://localhost:9000',
    http.Client? httpClient,
  }) : _http = httpClient ?? http.Client() {
    final trimmedBase = baseUrl.replaceAll(RegExp(r'/$'), '');
    connectors = ConnectorsManager(this, trimmedBase);
    tools = ToolsManager(this, trimmedBase);
    apiKeys = ApiKeysManager(this, trimmedBase);
  }

  Map<String, String> get _headers => {
        'Authorization': 'Bearer $apiKey',
        'Content-Type': 'application/json',
      };

  Future<dynamic> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? params,
  }) async {
    var url = '$baseUrl$path';
    if (params != null && params.isNotEmpty) {
      final queryString =
          params.entries.map((e) => '${Uri.encodeComponent(e.key)}=${Uri.encodeComponent(e.value)}').join('&');
      url = '$url?$queryString';
    }

    final uri = Uri.parse(url);
    http.Response resp;

    if (body != null) {
      resp = await _http
          .request(
            method,
            uri,
            headers: _headers,
            body: jsonEncode(body),
          )
          .timeout(const Duration(seconds: 30));
    } else {
      resp = await _http
          .request(method, uri, headers: _headers)
          .timeout(const Duration(seconds: 30));
    }

    if (resp.statusCode == 401) throw AuthException();
    if (resp.statusCode == 429) throw OminiConnectException('Rate limited', statusCode: 429);
    if (resp.statusCode == 404) throw OminiConnectException('Not found', statusCode: 404);
    if (resp.statusCode >= 400) {
      throw OminiConnectException(
        'Upstream error ${resp.statusCode}',
        statusCode: resp.statusCode,
        body: resp.body,
      );
    }

    if (resp.body.isEmpty) return null;
    return jsonDecode(resp.body);
  }

  Future<void> _requestNoContent(String method, String path, {Map<String, dynamic>? body}) async {
    final uri = Uri.parse('$baseUrl$path');
    final resp = await (_http.request(
      method,
      uri,
      headers: _headers,
      body: body != null ? jsonEncode(body) : null,
    )).timeout(const Duration(seconds: 30));

    if (resp.statusCode == 401) throw AuthException();
    if (resp.statusCode == 429) throw OminiConnectException('Rate limited', statusCode: 429);
    if (resp.statusCode >= 400) {
      throw OminiConnectException(
        'Upstream error ${resp.statusCode}',
        statusCode: resp.statusCode,
      );
    }
  }

  /// Maton-style direct API call — the simplest way to use OminiConnect.
  ///
  /// Example:
  /// ```dart
  /// final repos = await client.call('github', 'GET', '/user/repos', params: {'sort': 'updated'});
  /// await client.call('slack', 'POST', '/api/chat.postMessage', body: {'channel': 'C0123', 'text': 'Hi!'});
  /// ```
  Future<dynamic> call(
    String platform,
    String method,
    String path, {
    Map<String, String>? params,
    Map<String, dynamic>? body,
  }) async {
    final payload = <String, dynamic>{
      'method': method.toUpperCase(),
      'path': path,
    };
    if (params != null) payload['params'] = params;
    if (body != null) payload['body'] = body;

    final uri = Uri.parse('$baseUrl/api/call/$platform');
    final resp = await _http
        .post(uri, headers: _headers, body: jsonEncode(payload))
        .timeout(const Duration(seconds: 30));

    if (resp.statusCode == 401) throw AuthException();
    if (resp.statusCode == 429) throw OminiConnectException('Rate limited', statusCode: 429);
    if (resp.statusCode >= 400) {
      throw OminiConnectException(
        'Upstream error ${resp.statusCode}',
        statusCode: resp.statusCode,
        body: resp.body,
      );
    }

    if (resp.body.isEmpty) return null;
    return jsonDecode(resp.body);
  }
}

// ─── Connectors ─────────────────────────────────────────────────────────────

class ConnectorsManager {
  final OminiConnect _client;
  final String _base;

  ConnectorsManager(this._client, this._base);

  /// List all connected platforms.
  Future<List<Connector>> list() async {
    final data = await _client._request('GET', '/api/connectors');
    return (data as List<dynamic>)
        .map((c) => Connector.fromJson(c as Map<String, dynamic>))
        .toList();
  }

  /// Delete a connected platform.
  Future<void> delete(String platform) async {
    await _client._requestNoContent('DELETE', '/api/connectors/$platform');
  }
}

// ─── API Keys ───────────────────────────────────────────────────────────────

class ApiKeysManager {
  final OminiConnect _client;
  final String _base;

  ApiKeysManager(this._client, this._base);

  /// Create a new named API key.
  /// The raw key is returned ONLY here — store securely.
  Future<ApiKeyCreated> create(String label) async {
    final data = await _client._request('POST', '/auth/apikey', body: {'label': label});
    return ApiKeyCreated.fromJson(data as Map<String, dynamic>);
  }

  /// List all API keys (raw key is never returned).
  Future<List<ApiKeySummary>> list() async {
    final data = await _client._request('GET', '/auth/apikey');
    return (data as List<dynamic>)
        .map((k) => ApiKeySummary.fromJson(k as Map<String, dynamic>))
        .toList();
  }

  /// Revoke an API key.
  Future<void> delete(String keyHash) async {
    await _client._requestNoContent('DELETE', '/auth/apikey/$keyHash');
  }
}

// ─── Tools ─────────────────────────────────────────────────────────────────

class ToolsManager {
  final OminiConnect _client;
  final String _base;

  ToolsManager(this._client, this._base);

  /// List available tools, optionally filtered by platform.
  Future<ToolkitsResponse> list({String? platform}) async {
    final data = await _client._request(
      'GET',
      '/api/tools',
      params: platform != null ? {'platform': platform} : null,
    );
    return ToolkitsResponse.fromJson(data as Map<String, dynamic>);
  }

  /// Search tools by query string.
  Future<ToolsSearchResponse> search(
    String q, {
    String? platform,
    String? filterScope,
  }) async {
    final params = <String, String>{'q': q};
    if (platform != null) params['platform'] = platform;
    if (filterScope != null) params['filter_scope'] = filterScope;
    final data = await _client._request('GET', '/api/tools/search', params: params);
    return ToolsSearchResponse.fromJson(data as Map<String, dynamic>);
  }

  /// Execute a tool by slug with structured arguments.
  /// Set [callbackUrl] for async execution (returns immediately).
  Future<ToolExecuteResult> execute(
    String toolSlug,
    String platform, {
    Map<String, dynamic>? arguments,
    String? callbackUrl,
  }) async {
    final body = <String, dynamic>{
      'tool_slug': toolSlug,
      'platform': platform,
      'arguments': arguments ?? {},
    };
    if (callbackUrl != null) body['callback_url'] = callbackUrl;
    final data = await _client._request('POST', '/api/tools/execute', body: body);
    return ToolExecuteResult.fromJson(data as Map<String, dynamic>);
  }
}

File: sdk/ios/Sources/OminiConnect/OminiConnect.swift
// OminiConnect iOS SDK
// Connect AI agents to external platforms via OAuth

import Foundation

// ─── Errors ─────────────────────────────────────────────────────────────────

public enum OminiConnectError: Error, LocalizedError {
    case auth
    case connectorNotFound(String)
    case toolNotFound(String)
    case scopeInsufficient
    case upstream(statusCode: Int, body: String)
    case network(Error)
    case rateLimited

    public var errorDescription: String? {
        switch self {
        case .auth: return "Invalid or missing API key"
        case .connectorNotFound(let platform): return "Platform not connected: \(platform)"
        case .toolNotFound(let slug): return "Tool not found: \(slug)"
        case .scopeInsufficient: return "Insufficient scopes for this operation"
        case .upstream(let code, let body): return "Upstream error \(code): \(body)"
        case .network(let err): return "Network error: \(err.localizedDescription)"
        case .rateLimited: return "Rate limited — back off and retry"
        }
    }
}

// ─── Types ──────────────────────────────────────────────────────────────────

public struct Connector: Codable {
    public let platform: String
    public let enabled: Bool
    public let scopes: [String]
    public let createdAt: String

    enum CodingKeys: String, CodingKey {
        case platform, enabled, scopes
        case createdAt = "created_at"
    }

    public init(platform: String, enabled: Bool, scopes: [String], createdAt: String) {
        self.platform = platform
        self.enabled = enabled
        self.scopes = scopes
        self.createdAt = createdAt
    }
}

public struct ApiKeyCreated: Codable {
    public let key: String
    public let label: String
    public let createdAt: String

    enum CodingKeys: String, CodingKey {
        case key, label
        case createdAt = "created_at"
    }

    public init(key: String, label: String, createdAt: String) {
        self.key = key
        self.label = label
        self.createdAt = createdAt
    }
}

public struct ApiKeySummary: Codable {
    public let keyHash: String
    public let label: String
    public let createdAt: String

    enum CodingKeys: String, CodingKey {
        case keyHash = "key_hash"
        case label
        case createdAt = "created_at"
    }

    public init(keyHash: String, label: String, createdAt: String) {
        self.keyHash = keyHash
        self.label = label
        self.createdAt = createdAt
    }
}

public struct ToolSummary: Codable {
    public let slug: String
    public let name: String
    public let description: String
    public let method: String
    public let endpoint: String
    public let scopes: [String]
    public let scopeSatisfied: String
    public let tags: [String]

    enum CodingKeys: String, CodingKey {
        case slug, name, description, method, endpoint, scopes, tags
        case scopeSatisfied = "scope_satisfied"
    }

    public init(slug: String, name: String, description: String, method: String, endpoint: String, scopes: [String], scopeSatisfied: String, tags: [String]) {
        self.slug = slug
        self.name = name
        self.description = description
        self.method = method
        self.endpoint = endpoint
        self.scopes = scopes
        self.scopeSatisfied = scopeSatisfied
        self.tags = tags
    }
}

public struct Toolkit: Codable {
    public let slug: String
    public let name: String
    public let logo: String?
    public let provider: String
    public let tools: [ToolSummary]

    public init(slug: String, name: String, logo: String?, provider: String, tools: [ToolSummary]) {
        self.slug = slug
        self.name = name
        self.logo = logo
        self.provider = provider
        self.tools = tools
    }
}

public struct ToolkitsResponse: Codable {
    public let toolkits: [Toolkit]

    public init(toolkits: [Toolkit]) {
        self.toolkits = toolkits
    }
}

public struct ToolsSearchResponse: Codable {
    public let tools: [ToolSummary]
    public let query: String

    public init(tools: [ToolSummary], query: String) {
        self.tools = tools
        self.query = query
    }
}

public struct ToolExecuteResult: Codable {
    public let ok: Bool
    public let body: AnyCodableValue?
    public let error: String?
    public let callId: String?
    public let status: String?
    public let durationMs: Int?

    enum CodingKeys: String, CodingKey {
        case ok, body, error
        case callId = "call_id"
        case status
        case durationMs = "duration_ms"
    }

    public init(ok: Bool, body: AnyCodableValue?, error: String?, callId: String?, status: String?, durationMs: Int?) {
        self.ok = ok
        self.body = body
        self.error = error
        self.callId = callId
        self.status = status
        self.durationMs = durationMs
    }
}

/// Type-erased Codable value for flexible body types
public struct AnyCodableValue: Codable {
    public let value: Any

    public init(_ value: Any) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let intVal = try? container.decode(Int.self) {
            value = intVal
        } else if let doubleVal = try? container.decode(Double.self) {
            value = doubleVal
        } else if let stringVal = try? container.decode(String.self) {
            value = stringVal
        } else if let boolVal = try? container.decode(Bool.self) {
            value = boolVal
        } else if let arrayVal = try? container.decode([AnyCodableValue].self) {
            value = arrayVal.map { $0.value }
        } else if let dictVal = try? container.decode([String: AnyCodableValue].self) {
            value = dictVal.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let intVal = value as? Int {
            try container.encode(intVal)
        } else if let doubleVal = value as? Double {
            try container.encode(doubleVal)
        } else if let stringVal = value as? String {
            try container.encode(stringVal)
        } else if let boolVal = value as? Bool {
            try container.encode(boolVal)
        } else {
            try container.encodeNil()
        }
    }
}

// ─── Client ─────────────────────────────────────────────────────────────────

public class OminiConnect {
    private let baseURL: String
    private let apiKey: String
    private let session: URLSession

    public var connectors: ConnectorsManager!
    public var tools: ToolsManager!
    public var apiKeys: ApiKeysManager!
    public var llm: LlmManager!
    public var agents: AgentsManager!

    public init(apiKey: String, baseURL: String = "http://localhost:9000") {
        self.apiKey = apiKey
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)

        // Swift requires all stored properties initialized before self can be used.
        // Use nil placeholder initializers then assign via a helper.
        self.connectors = nil
        self.tools = nil
        self.apiKeys = nil
        self.llm = nil
        self.agents = nil

        // Now assign properly (self is fully initialized now)
        self.connectors = ConnectorsManager(client: self)
        self.tools = ToolsManager(client: self)
        self.apiKeys = ApiKeysManager(client: self)
        self.llm = LlmManager(client: self)
        self.agents = AgentsManager(client: self)
    }

    // ─── HTTP layer ───────────────────────────────────────────────────────────

    func request<T: Decodable>(
        _ method: String,
        _ path: String,
        body: Encodable? = nil,
        params: [String: String]? = nil
    ) async throws -> T {
        var urlComponents = URLComponents(string: baseURL + path)!
        if let params = params, !params.isEmpty {
            urlComponents.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
        }

        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = method
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let body = body {
            let encoder = JSONEncoder()
            request.httpBody = try encoder.encode(body)
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw OminiConnectError.network(URLError(.badServerResponse))
        }

        if httpResponse.statusCode == 401 {
            throw OminiConnectError.auth
        }
        if httpResponse.statusCode == 429 {
            throw OminiConnectError.rateLimited
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let bodyStr = String(data: data, encoding: .utf8) ?? ""
            throw OminiConnectError.upstream(statusCode: httpResponse.statusCode, body: bodyStr)
        }

        if httpResponse.statusCode == 204 || data.isEmpty {
            throw OminiConnectError.network(URLError(.cannotDecodeContentData))
        }

        let decoder = JSONDecoder()
        return try decoder.decode(T.self, from: data)
    }

    func requestNoContent(
        _ method: String,
        _ path: String,
        body: Encodable? = nil
    ) async throws {
        var urlComponents = URLComponents(string: baseURL + path)!

        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = method
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let body = body {
            let encoder = JSONEncoder()
            request.httpBody = try encoder.encode(body)
        }

        let (_, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw OminiConnectError.network(URLError(.badServerResponse))
        }

        if httpResponse.statusCode == 401 {
            throw OminiConnectError.auth
        }
        if httpResponse.statusCode == 429 {
            throw OminiConnectError.rateLimited
        }

        if !(200...299).contains(httpResponse.statusCode) {
            throw OminiConnectError.upstream(statusCode: httpResponse.statusCode, body: "")
        }
    }

    /// Maton-style direct API call — the simplest way to use OminiConnect.
    public func call(
        platform: String,
        method: String,
        path: String,
        params: [String: String]? = nil,
        body: [String: Any]? = nil
    ) async throws -> Any {
        var payload: [String: Any] = ["method": method.uppercased(), "path": path]
        if let params = params {
            payload["params"] = params
        }
        if let body = body {
            payload["body"] = body
        }

        let url = URL(string: "\(baseURL)/api/call/\(platform)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw OminiConnectError.network(URLError(.badServerResponse))
        }

        if httpResponse.statusCode == 401 {
            throw OminiConnectError.auth
        }
        if httpResponse.statusCode == 429 {
            throw OminiConnectError.rateLimited
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let bodyStr = String(data: data, encoding: .utf8) ?? ""
            throw OminiConnectError.upstream(statusCode: httpResponse.statusCode, body: bodyStr)
        }

        return try JSONSerialization.jsonObject(with: data)
    }
}

// ─── Connectors ───────────────────────────────────────────────────────────────

public class ConnectorsManager {
    private let client: OminiConnect

    init(client: OminiConnect) {
        self.client = client
    }

    /// List all connected platforms.
    public func list() async throws -> [Connector] {
        try await client.request("GET", "/api/connectors")
    }

    /// Delete a connected platform.
    public func delete(platform: String) async throws {
        try await client.requestNoContent("DELETE", "/api/connectors/\(platform)")
    }
}

// ─── API Keys ────────────────────────────────────────────────────────────────

private struct CreateApiKeyRequest: Encodable {
    let label: String
}

public class ApiKeysManager {
    private let client: OminiConnect

    init(client: OminiConnect) {
        self.client = client
    }

    /// Create a new named API key. The raw key is returned ONLY here — store securely.
    public func create(label: String) async throws -> ApiKeyCreated {
        try await client.request("POST", "/auth/apikey", body: CreateApiKeyRequest(label: label))
    }

    /// List all API keys (raw key is never returned).
    public func list() async throws -> [ApiKeySummary] {
        try await client.request("GET", "/auth/apikey")
    }

    /// Revoke an API key.
    public func delete(keyHash: String) async throws {
        try await client.requestNoContent("DELETE", "/auth/apikey/\(keyHash)")
    }
}

// ─── Tools ──────────────────────────────────────────────────────────────────

public class ToolsManager {
    private let client: OminiConnect

    init(client: OminiConnect) {
        self.client = client
    }

    /// List available tools, optionally filtered by platform.
    public func list(platform: String? = nil) async throws -> ToolkitsResponse {
        var params: [String: String]? = nil
        if let p = platform {
            params = ["platform": p]
        }
        return try await client.request("GET", "/api/tools", params: params)
    }

    /// Search tools by query string.
    public func search(q: String, platform: String? = nil, filterScope: String? = nil) async throws -> ToolsSearchResponse {
        var params: [String: String] = ["q": q]
        if let p = platform { params["platform"] = p }
        if let f = filterScope { params["filter_scope"] = f }
        return try await client.request("GET", "/api/tools/search", params: params)
    }

    /// Execute a tool by slug with structured arguments.
    /// Set callbackURL for async execution (returns immediately).
    public func execute(
        toolSlug: String,
        platform: String,
        arguments: [String: Any]? = nil,
        callbackURL: String? = nil
    ) async throws -> ToolExecuteResult {
        struct ExecuteRequest: Encodable {
            let toolSlug: String
            let platform: String
            let arguments: [String: AnyCodable]
            let callbackURL: String?

            enum CodingKeys: String, CodingKey {
                case toolSlug = "tool_slug"
                case platform
                case arguments
                case callbackURL = "callback_url"
            }
        }

        var argsDict: [String: AnyCodable] = [:]
        if let arguments = arguments {
            for (k, v) in arguments {
                argsDict[k] = AnyCodable(v)
            }
        }

        let payload = ExecuteRequest(
            toolSlug: toolSlug,
            platform: platform,
            arguments: argsDict,
            callbackURL: callbackURL
        )

        return try await client.request("POST", "/api/tools/execute", body: payload)
    }
}

/// Helper for encoding arbitrary values
private struct AnyCodable: Encodable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let intVal = value as? Int {
            try container.encode(intVal)
        } else if let doubleVal = value as? Double {
            try container.encode(doubleVal)
        } else if let stringVal = value as? String {
            try container.encode(stringVal)
        } else if let boolVal = value as? Bool {
            try container.encode(boolVal)
        } else {
            try container.encodeNil()
        }
    }
}
